File: 프로그래머스/3/49191. 순위/순위.py
from collections import deque
graphA = [[]]
graphB = [[]]
def solution(n, results):
    global graphA, graphB
    
    graphA = [[] for _ in range(n+1)]
    graphB = [[] for _ in range(n+1)]
    
    for s, e in results:
        graphA[s].append(e)
        graphB[e].append(s)
        
    cnt = 0
    for i in range(1, n+1):
        visited = [False] * (n+1)
        bfs(i, graphA, visited)
        bfs(i, graphB, visited)
        
        if check(visited, n):
            cnt += 1
    return cnt

def bfs(start, graph,visited):
    q = deque()
    q.append(start)
    visited[start] = True
    
    while q:
        now = q.popleft()
        
        for nxt in graph[now]:
            if not visited[nxt]:
                visited[nxt] = True
                q.append(nxt)
                
def check(visited, n):
    for i in range(1, n+1):
        if not visited[i]:
            return False
    return True
        
            

File: 프로그래머스/3/49191. 순위/순위.java
import java.util.*;

class Solution {
    static ArrayList<Integer>[] winList;
    static ArrayList<Integer>[] loseList;

    public int solution(int n, int[][] results) {
        winList = new ArrayList[n + 1];
        loseList = new ArrayList[n + 1];

        for (int i = 1; i <= n; i++) {
            winList[i] = new ArrayList<>();
            loseList[i] = new ArrayList<>();
        }

        for (int[] result : results) {
            int winner = result[0];
            int loser = result[1];

            winList[winner].add(loser);
            loseList[loser].add(winner);
        }

        int answer = 0;

        for (int i = 1; i <= n; i++) {
            int winCount = bfs(i, winList, n);
            int loseCount = bfs(i, loseList, n);

            if (winCount + loseCount == n - 1) {
                answer++;
            }
        }

        return answer;
    }

    static int bfs(int start, ArrayList<Integer>[] graph, int n) {
        ArrayDeque<Integer> queue = new ArrayDeque<>();
        boolean[] visited = new boolean[n + 1];

        visited[start] = true;
        queue.add(start);

        int count = 0;

        while (!queue.isEmpty()) {
            int now = queue.poll();

            for (int next : graph[now]) {
                if (!visited[next]) {
                    visited[next] = true;
                    queue.add(next);
                    count++;
                }
            }
        }

        return count;
    }
}